src/compiler/checker.ts

```diff
@@ -29393,9 +29393,9 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
 
         const isAlias = localOrExportSymbol.flags & SymbolFlags.Alias;
 
-        // We only narrow variables and parameters occurring in a non-assignment position. For all other
+        // We only narrow variables, parameters and functions occurring in a non-assignment position. For all other
         // entities we simply return the declared type.

```

Comment:
tbf... if this is something the plan would like to pull in, we probably should enable narrowing on just anything. `asserts` make it possible to expand all non-nullable types